### ISSUE TITLE
"Channel has been shut down" errors in automation.

### ISSUE THREAD

**Issue by tmccombs** [pulumi/pulumi#6998]
## Expected behavior
Not getting spammed with a bunch of difficult-to-understand errors.

## Current behavior
Getting a bunch of errors like:

```
(node:483560) UnhandledPromiseRejectionWarning: Error: Channel has been shut down
    at ChannelImplementation.createCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/channel.ts:608:13)
    at getCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:319:24)
    at getBottomInterceptingCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:447:16)
    at /home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:541:7
    at Object.getInterceptingCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:543:10)
    at ServiceClientImpl.makeUnaryRequest (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client.ts:299:45)
    at ServiceClientImpl.registerResource (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/make-client.ts:186:15)
    at /home/thayne/sandbox/pulumi-test/node_modules/@pulumi/pulumi/runtime/resource.js:268:106
    at new Promise (<anonymous>)
    at Object.<anonymous> (/home/thayne/sandbox/pulumi-test/node_modules/@pulumi/pulumi/runtime/resource.js:268:65)
    at emitUnhandledRejectionWarning (internal/process/promises.js:168:15)
    at processPromiseRejections (internal/process/promises.js:247:11)
    at processTicksAndRejections (internal/process/task_queues.js:94:32)
(node:483560) Error: Channel has been shut down
    at ChannelImplementation.createCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/channel.ts:608:13)
    at getCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:319:24)
    at getBottomInterceptingCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:447:16)
    at /home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:541:7
    at Object.getInterceptingCall (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client-interceptors.ts:543:10)
    at ServiceClientImpl.makeUnaryRequest (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/client.ts:299:45)
    at ServiceClientImpl.registerResource (/home/thayne/sandbox/pulumi-test/node_modules/@grpc/grpc-js/src/make-client.ts:186:15)
    at /home/thayne/sandbox/pulumi-test/node_modules/@pulumi/pulumi/runtime/resource.js:268:106
    at new Promise (<anonymous>)
    at Object.<anonymous> (/home/thayne/sandbox/pulumi-test/node_modules/@pulumi/pulumi/runtime/resource.js:268:65)
```

when running an automation program

## Steps to reproduce

1. Use a program like:
```
import { LocalWorkspace } from "@pulumi/pulumi/automation";
import * as tls from "@pulumi/tls";

async function program() {
  const privateKey = new tls.PrivateKey("test", {
    algorithm: "ECDSA",
    ecdsaCurve: "P384",
  });

  const ca = new tls.SelfSignedCert("myca", {
    privateKeyPem: privateKey.privateKeyPem,
    keyAlgorithm: privateKey.algorithm,
    allowedUses: ["cert_signing", "crl_signing"],
    validityPeriodHours: 1000 * 3600 * 24,
    subjects: [
      {
        commonName: "test.example.com",
        country: "US",
        locality: "Utah",
        organization: "Example",
      },
    ],
    isCaCertificate: true,
  });
}

async function run() {
  console.info("creating stack");
  const stack = await LocalWorkspace.createOrSelectStack({
    stackName: "dev",
    projectName: "test",
    program: program,
  });

  let command = "preview";
  if (process.argv.length > 2) {
    command = process.argv[2];
  }

  if (command == "preview") {
    console.log("Preview:");
    const result = await stack.preview({
      diff: true,
      onOutput: console.info,
    });
  } else if (command == "up") {
    const result = await stack.up({
      onOutput: console.info,
    });
  } else {
    console.error(`Unknown command ${command}`);
  }
  console.info("done");
}

run().catch((err) => console.error(err));

``` 
with a package.json like: 
```
{
    "name": "pulumi-test",
    "main": "index.ts",
    "dependencies": {
        "@pulumi/pulumi": "^3.0.0",
        "@pulumi/tls": "^4.0.0",
        "@types/node": "^10.0.0",
        "ts-node": "^9.1.1"
    },
    "scripts": {
        "start": "./node_modules/ts-node/dist/bin.js index.ts"
    }
}

```

2. run `npm run start up`

## Context (Environment)
I just ran into this while experimenting with the automation API. It looks like the rest of the program works as intended but this output makes the output very difficult to read. I also don't know what the error means, so I'm not sure if there are other negative side effects. I'm using the local backend while experimenting.

Interestingly, if I comment out the SelfSignedCert resource, then I don't get the error.

## Affected feature
automation API

**Comment by lblackstone** [comment]
@EvanBoyle I'm not sure if you're already tracking this issue elsewhere?

**Comment by tmccombs** [comment]
I was able to get the errors to stop if I add a `return {};` to the end of my program. So maybe that is my mistake, although the errors definitely didn't lead me in the right direction.

**Comment by roderik** [comment]
FYI, hitting the same

<img width="1032" alt="CleanShot 2021-05-22 at 15 19 49@2x" src="https://user-images.githubusercontent.com/16780/119228022-37352500-bb11-11eb-9748-a694b92eec2b.png">

To make this even weirder, the "program" we run is the same for GKE, AKS and EKS. And it is a different stack that exposes the kubeconfigs of these three clusters. It fails this way, more like "hangs" because it keeps running, for AKS (anecdotal from a colleague) and EKS, but works fine for GKE.

Why mentioning anecdotal, is that the eks one is the only one that is not marked as a "secret", which might be relevant.

**Comment by EvanBoyle** [comment]
This is an opaque error from grpc: https://github.com/grpc/grpc-node/blob/master/packages/grpc-js/src/channel.ts#L575-L577

Curious @roderik, are there any differences in runtime between the three stacks?

**Comment by roderik** [comment]
No, 100% the same, we choose "clusters" based on the name which is also the name of the exported values. So eks-europe, aks-europe, gke-europe is th eonly difference between the runs. 

Not that this runs in a queuing system, inside a GKE cluster, deploying to other clusters, from automation based code, so making sure the right env vars are available and correct is a challenge.

For example,

```

    await stack.setAllConfig({
      'cloudflare:accountId': {
        value: this.configService.get<string>('cloudflareAccountId'),
      },
      'cloudflare:apiKey': {
        value: this.configService.get<string>('cloudflareApiKey'),
        secret: true,
      },
      'cloudflare:email': {
        value: this.configService.get<string>('cloudflareEmail'),
      },
      'gcp:project': {
        value: this.configService.get<string>('googleCloudProjectId'),
      },
      'gcp:zone': { value: 'europe-west1-b' },
      'gcp:credentials': {
        value: JSON.stringify({
          type: 'service_account',
          project_id: this.configService.get<string>('googleCloudProjectId'),
          private_key_id: this.configService.get<string>('googleCloudPrivateKeyId'),
          private_key: this.configService.get<string>('googleCloudPrivateKey').replace(/\\n/g, '\n'),
          client_email: this.configService.get<string>('googleCloudClientEmail'),
          client_id: this.configService.get<string>('googleCloudClientId'),
          auth_uri: 'https://accounts.google.com/o/oauth2/auth',
          token_uri: 'https://oauth2.googleapis.com/token',
          auth_provider_x509_cert_url: 'https://www.googleapis.com/oauth2/v1/certs',
          client_x509_cert_url: this.configService.get<string>('googleCloudClientX509CertUrl'),
        }),
        secret: true,
      },
      'azure:environment': { value: 'public' },
      'azure:location': { value: 'westeurope' },
      'azure:clientId': {
        value: this.configService.get<string>('azureApplicationId'),
      },
      'azure:clientSecret': {
        value: this.configService.get<string>('azureClientSecret'),
        secret: true,
      },
      'azure:subscriptionId': {
        value: this.configService.get<string>('azureSubscriptionId'),
      },
      'azure:tenantId': {
        value: this.configService.get<string>('azureTenant'),
      },
      'aws:region': { value: 'eu-west-3' },
      'aws:accessKey': {
        value: this.configService.get<string>('awsAccessKeyId'),
      },
      'aws:secretKey': {
        value: this.configService.get<string>('awsSecretAccessKey'),
        secret: true,
      },
    });
```

is not enough, we also need to set the GOOGLE_CREDENTIALS env vans to the full credentials json.

Now, this is built up over time and over many pulumi versions, maybe it is not needed anymore but this works.

For azure and was this used to work, but i'm starting to think that i need env vars s well to make this work.

**Comment by roderik** [comment]
After "hanging" for an entire night I killed the pod and on restart it started fresh, giving a new error, after doing "more" tasks on the cluster.

```
/usr/src/app/node_modules/@grpc/grpc-js/src/call.ts:81                                                                                                                                                                                                             
  return Object.assign(new Error(message), status);                                                                                                                                                                                                                
                       ^                                                                                                                                                                                                                                           
Error: Resource monitor is terminating                                                                                                                                                                                                                             
    at Object.callErrorFromStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/call.ts:81:24)                                                                                                                                                                      
    at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client.ts:338:36)                                                                                                                                                                       
    at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:426:34)                                                                                                                                                          
    at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:389:48)                                                                                                                                                          
    at /usr/src/app/node_modules/@grpc/grpc-js/src/call-stream.ts:276:24                                                                                                                                                                                           
    at processTicksAndRejections (node:internal/process/task_queues:78:11)   
```

**Comment by roderik** [comment]
When it stalls, these are the processes

<img width="1924" alt="CleanShot 2021-05-23 at 08 57 48@2x" src="https://user-images.githubusercontent.com/16780/119251123-0ef70600-bba5-11eb-9cc9-04212fd6c5b8.png">

The event log it mentions contains (cut off the lines because it is full of secrets), but I see the outputs I need from the stackreference in sequence 29 and 30

```
{"sequence":26,"timestamp":1621753027,"resourcePreEvent":{"metadata":{"op":"read","urn":"urn:pulumi:luis-test-f-9c::bpaas::pulumi:pulumi:StackReference::settlemint/bpaas-clusters/production","type":"pulumi:pulumi:StackReference","old":null,"new":
{"sequence":27,"timestamp":1621753027,"diagnosticEvent":{"prefix":"\u003c{%reset%}\u003edebug: \u003c{%reset%}\u003e","message":"\u003c{%reset%}\u003eRegisterResourceOutputs RPC finished: urn=urn:pulumi:luis-test-f-9c::bpaas::bpaas:cluster-serv
{"sequence":28,"timestamp":1621753027,"diagnosticEvent":{"prefix":"\u003c{%reset%}\u003edebug: \u003c{%reset%}\u003e","message":"\u003c{%reset%}\u003eRegisterResourceOutputs RPC finished: urn=urn:pulumi:luis-test-f-9c::bpaas::bpaas:cluster-serv
{"sequence":29,"timestamp":1621753036,"resOutputsEvent":{"metadata":{"op":"read","urn":"urn:pulumi:luis-test-f-9c::bpaas::pulumi:pulumi:StackReference::settlemint/bpaas-clusters/production","type":"pulumi:pulumi:StackReference","old":null,"new"
{"sequence":30,"timestamp":1621753037,"diagnosticEvent":{"prefix":"\u003c{%reset%}\u003edebug: \u003c{%reset%}\u003e","message":"\u003c{%reset%}\u003eReadResource RPC finished: resource:settlemint/bpaas-clusters/production[pulumi:pulumi:StackReferenc
{"sequence":31,"timestamp":1621753037,"diagnosticEvent":{"prefix":"\u003c{%fg 1%}\u003eerror: \u003c{%reset%}\u003e","message":"\u003c{%reset%}\u003eUnhandled exception: Error: Channel has been shut down\n    at ChannelImplementation.createCall (/usr/src/app/node_modules/@grpc/grpc-js/src/channel.ts:608:13)\n    at getCall (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:319:24)\n    at getBottomInterceptingCall (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:447:16)\n    at /usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:541:7\n    at Object.getInterceptingCall (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:543:10)\n    at ServiceClientImpl.makeUnaryRequest (/usr/src/app/node_modules/@grpc/grpc-js/src/client.ts:303:45)\n    at ServiceClientImpl.registerResource (/usr/src/app/node_modules/@grpc/grpc-js/src/make-client.ts:186:15)\n    at /usr/src/app/node_modules/@pulumi/kubernetes/node_modules/@pulumi/pulumi/runtime/resource.js:268:106\n    at new Promise (\u003canonymous\u003e)\n    at Object.\u003canonymous\u003e (/usr/src/app/node_modules/@pulumi/kubernetes/node_modules/@pulumi/pulumi/runtime/resource.js:268:65)\u003c{%reset%}\u003e\n","color":"raw","severity":"error"}}
```

**Comment by roderik** [comment]
I thought Node 16 might be the culprit, but it also fails with Node 14

```
│ (node:1) UnhandledPromiseRejectionWarning: Error: Resource monitor is terminating                                                                                                                                                                            │
│     at Object.callErrorFromStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/call.ts:81:24)                                                                                                                                                                │
│     at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client.ts:338:36)                                                                                                                                                                 │
│     at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:426:34)                                                                                                                                                    │
│     at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:389:48)                                                                                                                                                    │
│     at /usr/src/app/node_modules/@grpc/grpc-js/src/call-stream.ts:276:24                                                                                                                                                                                     │
│     at processTicksAndRejections (internal/process/task_queues.js:77:11)                                                                                                                                                                                     │
│ (Use `node --trace-warnings ...` to show where the warning was created)   
```

**Comment by roderik** [comment]
Downgrading to Pulumi 3.2.1 (cli) did not help

**Comment by roderik** [comment]
@EvanBoyle I can confirm this is a regression in v3.3.0. Using Node 16, cli 3.2.1 and sdk 3.2.1 it works flawlessly every time on all clouds. Last change was downgrading the SDK, so I assume the issue is in the SDK part

**Comment by EvanBoyle** [comment]
@roderik there were two changes to the node SDK in the last release, would you mind trying your repro with the dev channel builds to help bisect the issue? 

- https://www.npmjs.com/package/@pulumi/pulumi/v/3.3.0-alpha.1621034643
- https://www.npmjs.com/package/@pulumi/pulumi/v/3.3.0-alpha.1620355411

**Comment by roderik** [comment]
- 3.3.0-alpha.1621034643 with cli 3.2.1 and with cli 3.3.0 has the following issue
- 3.3.0-alpha.1620355411 with cli 3.2.1 and with cli 3.3.0 also has the following issue
- 3.2.1 with cli 3.2.1 works

While it errors out, it does appear as "successful" in the pulumi site. 
My container however dies since it throws unhandled exception (even though we wrap everything in try catches)

```
│ /usr/src/app/node_modules/@grpc/grpc-js/src/call.ts:81                                                                                                                                                                                                       │
│   return Object.assign(new Error(message), status);                                                                                                                                                                                                          │
│                        ^                                                                                                                                                                                                                                     │
│ Error: Resource monitor is terminating                                                                                                                                                                                                                       │
│     at Object.callErrorFromStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/call.ts:81:24)                                                                                                                                                                │
│     at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client.ts:338:36)                                                                                                                                                                 │
│     at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:426:34)                                                                                                                                                    │
│     at Object.onReceiveStatus (/usr/src/app/node_modules/@grpc/grpc-js/src/client-interceptors.ts:389:48)                                                                                                                                                    │
│     at /usr/src/app/node_modules/@grpc/grpc-js/src/call-stream.ts:276:24                                                                                                                                                                                     │
│     at processTicksAndRejections (node:internal/process/task_queues:78:11) 
```

**Comment by EvanBoyle** [comment]
There don't appear to be any other SDK changes (other than to testing) between 3.2.1 and the pre-release build 3.3.0-alpha.1620355411. This could be a red herring, or perhaps some subtle difference in the npm package set. 

If the pulumi engine is observing the deployment as successfully completed, then it sounds like there may be a dependency in the graph that isn't fully captured by the engine (as also seems to be the case in the parent issue). Are you creating any resources within an apply, or using pulumi kubernetes helm support? 

Alternatively, do you have any "dangling resources" that are not used as inputs to other resources or not exported as stack exports at all? It would be worth trying to export any resources as a workaround: https://github.com/pulumi/pulumi/issues/6998#issuecomment-838989885

**Comment by tmccombs** [comment]
> do you have any "dangling resources" that are not used as inputs to other resources or not exported as stack exports at all?

That was the case for me.

I also, kind of suspect this is a race condition. With problematic config, it would fail most of the time, but if I remember correctly, sometimes it would succeed.

**Comment by roderik** [comment]
it is full of dangling resources, we create namespaces, k8s jobs etc etc, but we do not export them because we do not need them outside of the script. In this specific case we also deploy an empty k8s namespace we do not use (depends on user input if we use it or not). It seems wrong that I need to export everything to bypass this error

**Comment by EvanBoyle** [comment]
> It seems wrong that I need to export everything to bypass this error
I agree there is a deeper underlying issue to fix here. Just trying to see if we can find you an immediate workaround.

**Comment by roderik** [comment]
I don't need 3.3 at the moment so fine to stay on 3.2.1 for now, but you guys are moving so fast with cool new stuff, that it is not a 3 month solution :)

**Comment by roderik** [comment]
Is there any way we can help in getting this fixed?

We still have crashed now and again

```
/usr/src/app/node_modules/@pulumi/pulumi/runtime/resource.js:219                                                                                                                                                                                                                                                                                                            
    const preallocError = new Error();                                                                                                                                                                                                                                                                                                                                      
                          ^                                                                                                                                                                                                                                                                                                                                                 
Error: failed to register new resource networkp10-e0/NetworkP10 [kubernetes:core/v1:ServiceAccount]: Resource monitor is terminating                                                                                                                                                                                                                                        
    at Object.registerResource (/usr/src/app/node_modules/@pulumi/pulumi/runtime/resource.js:219:27)                                                                                                                                                                                                                                                                        
    at new Resource (/usr/src/app/node_modules/@pulumi/pulumi/resource.js:215:24)                                                                                                                                                                                                                                                                                           
    at new CustomResource (/usr/src/app/node_modules/@pulumi/pulumi/resource.js:307:9)                                                                                                                                                                                                                                                                                      
    at new ServiceAccount (/usr/src/app/node_modules/@pulumi/core/v1/serviceAccount.ts:91:9)                                                                                                                                                                                                                                                                                
    at /usr/src/app/node_modules/@pulumi/yaml/yaml.ts:3376:27                                                                                                                                                                                                                                                                                                               
    at /usr/src/app/node_modules/@pulumi/pulumi/output.js:250:35                                                                                                                                                                                                                                                                                                            
    at Generator.next (<anonymous>)                                                                                                                                                                                                                                                                                                                                         
    at /usr/src/app/node_modules/@pulumi/pulumi/output.js:21:71                                                                                                                                                                                                                                                                                                             
    at new Promise (<anonymous>)                                                                                                                                                                                                                                                                                                                                            
    at __awaiter (/usr/src/app/node_modules/@pulumi/pulumi/output.js:17:12) 
```

**Comment by roderik** [comment]
@EvanBoyle ok, i can confirm that 2.5.1 also does not work, nor 3.4.0.

It feels like there is some race condition. In a few runs of the same automation script, same settings, different "name". This deploys a new stack, with a namespace, cloudflare url + a helm chart, all based on that name passed in.

- first run, did not work. It seemed to get to the end and then stalled for a long time, ending up in the Resource monitor is terminating error. It left some stuff running

<img width="1594" alt="CleanShot 2021-06-07 at 11 01 14@2x" src="https://user-images.githubusercontent.com/16780/120989383-c84a0400-c77f-11eb-971a-67ecf3fe01be.png">

- next run, completed successfully

- next run, (i killed the running processes before), failed with an "invoke" type error.  

<img width="1272" alt="CleanShot 2021-06-07 at 10 58 37@2x" src="https://user-images.githubusercontent.com/16780/120989290-abadcc00-c77f-11eb-810a-a63cc691e08f.png">

- next run, failed to create the namespace before doing other things

``` 
+  bpaas:chart:corda-nms corda-six-6d creating error: post-step event returned an error: failed to verify snapshot: child resource urn:pulumi:development-corda-six-6d::bpaas::bpaas:blockchain-network$kubernetes:core/v1:Namespace$bpaas:chart:corda-nms::corda-six-6d refers to missing parent urn:pulumi:development-corda-six-6d::bpaas::bpaas:blockchain-network$kubernetes:core/v1:Namespace::
```

So it appears to be completely random, with different failures all around.

**Comment by EvanBoyle** [comment]
@roderik sorry to hear this is continuing to cause trouble. Two questions for you:

1. I notice you're using node 16. We haven't yet brought this version into our testing matrix. Does the problem reproduce on node 14?
2. Is it possible that your automation script saves or caches references to resources? Often these `failed to verify snapshot` errors can indicate that an automation API program saves a resource created in one run, and then references it in a later run. Resources can't be cached between program runs as calling `new Resource(...)` adds a reference to the dependency graph. If `new` isn't called, then the pulumi engine will thing the resource is missing and/or deleted and any resources that depend on that resource will fail.

**Comment by roderik** [comment]
1. Node 14 does not work on ARM macs which makes it really hard to develop locally 
2. There is no caching

I found something which might or might not be related. In the run we deploy 3 k8s jobs. I noticed that they were very slow. It feels like something times out waiting for the job to be finished. The job is successful at some point. I mention it because this job was not slow before.

**Comment by tmccombs** [comment]
When I ran into it, I was using node 14.

**Comment by lukehoban** [comment]
@roderik @EvanBoyle am I correct that the issue here was caused by having multiple copies of `@pulumi/pulumi` in use (one as a nested dependency of `@pulumi/kubernetes`)?  And that these issues are resolved when you ensure that a single version of that library is installed?

Addressing the underlying issues in automation api that lead to multiple instances of Pulumi/Pulumi triggering errors like this is I believe tracked in https://github.com/pulumi/pulumi/issues/5449 and https://github.com/pulumi/pulumi/issues/7276.

**Comment by tmccombs** [comment]
No, I only have a single `@pulumi/pulumi` in my node_modules.

**Comment by lukehoban** [comment]
> No, I only have a single @pulumi/pulumi in my node_modules.

Just to make doubly sure - since I believe we had been pretty convinced this was due to duplicate copies of `@pulumi/pulumi` - could you confirm that your `package-lock.json` has only a single instance of `@pulumi/pulumi` in it? 

> Interestingly, if I comment out the SelfSignedCert resource, then I don't get the error.

This also is a strong hint that this would be due to loading in a second instance of `@pulumi/pulumi`.  Bu removing that line of code, the import would most likely get elided by TypeScript, so the import of `tls` into Node would not happen, which would skip importing a potentially different version of the `@pulumi/pulumi` dependency via that indirect dependency.  Not a sure thing - but that data point is surprisingly consistent with this assumed root cause - hence why I'd love to be sure.

**Comment by tmccombs** [comment]
```
❯ grep pulumi package-lock.json
    "name": "pulumi-test",
        "@pulumi/pulumi": {
            "resolved": "https://registry.npmjs.org/@pulumi/pulumi/-/pulumi-3.2.0.tgz",
                "@pulumi/query": "^0.3.0",
        "@pulumi/query": {
            "resolved": "https://registry.npmjs.org/@pulumi/query/-/query-0.3.0.tgz",
        "@pulumi/tls": {
            "resolved": "https://registry.npmjs.org/@pulumi/tls/-/tls-4.0.0.tgz",
                "@pulumi/pulumi": "^3.0.0-alpha.0"
 ```
 
 ```
 ❯ find -name pulumi
./node_modules/@pulumi/pulumi
```

**Comment by EvanBoyle** [comment]
@tmccombs we've released a new version of the nodejs SDK, would you mind updating your repro to use it and let us know if the behavior has changed at all? https://www.npmjs.com/package/@pulumi/pulumi/v/3.6.0

**Comment by tmccombs** [comment]
I'm still getting the same error.

**Comment by EvanBoyle** [comment]
Repro'd this and it looks like there is a dropped dependency somewhere even with the latest node SDK. Will need to dig deeper to understand why, but for the time being @tmccombs exporting some property of the certificate to explicity inject that resource into the output dependency graph is an effective workaround:

```typescript
  return { ca: ca.id }
```

**Comment by EvanBoyle** [comment]
Ok, I did some debugging and determined that the problem is that the program exits "too quickly" and the shutdown routine runs before any of the async work wind up in the rpcKeepAlive queue. 

If we pump the event loop once before running the shutdown routine, the program runs as expected without modification:

https://github.com/pulumi/pulumi/compare/evan/nodeAutoRace

Need to do some more consideration on whether or not this is the fix we want.

**Comment by EvanBoyle** [comment]
@tmccombs you should be able to pull in the latest alpha with the fix (official 3.10 will come next week): https://www.npmjs.com/package/@pulumi/pulumi/v/3.10.0-alpha.1628106956

**Comment by tmccombs** [comment]
Yes, that seemed to fix it.
